File: test-suite/src/test/java/io/micronaut/security/docs/reactiveauthenticationprovider/CustomAuthenticationProvider.java
package io.micronaut.security.docs.reactiveauthenticationprovider;

import io.micronaut.context.annotation.Requires;
import io.micronaut.core.async.annotation.SingleResult;
import io.micronaut.http.HttpRequest;
import io.micronaut.security.authentication.AuthenticationFailureReason;
import io.micronaut.security.authentication.AuthenticationRequest;
import io.micronaut.security.authentication.AuthenticationResponse;
import io.micronaut.security.authentication.provider.HttpRequestReactiveAuthenticationProvider;
import jakarta.inject.Singleton;
import org.reactivestreams.Publisher;
import reactor.core.publisher.Mono;

@Requires(property = "spec.name", value = "ReactiveAuthenticationProviderTest")
//tag::clazz[]
@Singleton
class CustomAuthenticationProvider<B> implements HttpRequestReactiveAuthenticationProvider<B> {
    @Override
    @SingleResult
    public Publisher<AuthenticationResponse> authenticate(HttpRequest<B> requestContext, AuthenticationRequest<String, String> authRequest) {
        AuthenticationResponse rsp = authRequest.getIdentity().equals("user") && authRequest.getSecret().equals("password")
                ? AuthenticationResponse.success("user")
                : AuthenticationResponse.failure(AuthenticationFailureReason.CREDENTIALS_DO_NOT_MATCH);
        return Mono.create(emitter -> {
            emitter.success(rsp);
        });
    }
}
//end::clazz[]

File: test-suite-python/src/test/python/micronaut/security/docs/reactiveauthenticationprovider/CustomAuthenticationProvider.py
from jakarta.inject import Singleton
from micronaut.context.annotation import Requires
from micronaut.core.async_.annotation import SingleResult
from micronaut.http import HttpRequest
from micronaut.security.authentication import AuthenticationFailureReason, AuthenticationRequest, AuthenticationResponse
from micronaut.security.authentication.provider import HttpRequestReactiveAuthenticationProvider
from org.reactivestreams import Publisher
from reactor.core.publisher import Mono


@Requires(property="spec.name", value="ReactiveAuthenticationProviderTest")
# tag::clazz[]
@Singleton
class CustomAuthenticationProvider(HttpRequestReactiveAuthenticationProvider):

    @SingleResult
    def authenticate(self, requestContext: HttpRequest, authRequest: AuthenticationRequest) -> Publisher[AuthenticationResponse]:
        if authRequest.getIdentity() == "user" and authRequest.getSecret() == "password":
            rsp = AuthenticationResponse.success("user")
        else:
            rsp = AuthenticationResponse.failure(AuthenticationFailureReason.CREDENTIALS_DO_NOT_MATCH)
        return Mono.create(lambda emitter: emitter.success(rsp))
# end::clazz[]
